### PR TITLE
Paged search ordering revision

### PR DESCRIPTION
**Pull request by AliveMen**
added thenById

### CODE REVIEW

File: VirtoCommerce.MarketingModule.Data/Services/CouponService.cs [hunk 1]
```diff
@@ -51,7 +51,7 @@ public GenericSearchResult<Coupon> SearchCoupons(CouponSearchCriteria criteria)
                 {
                     sortInfos = new[] { new SortInfo { SortColumn = ReflectionUtility.GetPropertyName<Coupon>(x => x.Code), SortDirection = SortDirection.Descending } };
                 }
-                query = query.OrderBySortInfos(sortInfos);
+                query = query.OrderBySortInfos(sortInfos).ThenBy(x=>x.Id);
```
Comment (artem-dudarev) [inline VirtoCommerce.MarketingModule.Data/Services/CouponService.cs]: Invalid formatting. Should be `x => x.Id`

File: VirtoCommerce.MarketingModule.Data/Services/PromotionUsageService.cs [hunk 2]
```diff
@@ -38,7 +38,7 @@ public virtual GenericSearchResult<PromotionUsage> SearchUsages(PromotionUsageSe
                 {
                     sortInfos = new[] { new SortInfo { SortColumn = ReflectionUtility.GetPropertyName<PromotionUsage>(x => x.ModifiedDate), SortDirection = SortDirection.Descending } };
                 }
-                query = query.OrderBySortInfos(sortInfos);
+                query = query.OrderBySortInfos(sortInfos).ThenBy(x=>x.Id);
```
Comment (artem-dudarev) [inline VirtoCommerce.MarketingModule.Data/Services/PromotionUsageService.cs]: Invalid formatting. Should be `x => x.Id`
Turn on `Format document on save`

File: VirtoCommerce.MarketingModule.Data/Services/MarketingSearchServiceImpl.cs [hunk 3]
```diff
@@ -36,7 +36,7 @@ public virtual GenericSearchResult<coreModel.Promotion> SearchPromotions(Promoti
                 {
                     sortInfos = new[] { new SortInfo { SortColumn = ReflectionUtility.GetPropertyName<coreModel.Promotion>(x => x.Priority), SortDirection = SortDirection.Descending } };
                 }
-                query = query.OrderBySortInfos(sortInfos);
+                query = query.OrderBySortInfos(sortInfos).ThenBy(x=>x.Id);
```
Comment (artem-dudarev) [inline VirtoCommerce.MarketingModule.Data/Services/MarketingSearchServiceImpl.cs]: I guess I have to point to EVERY file in the PR to fix the same issue.

File: VirtoCommerce.MarketingModule.Data/Services/CouponService.cs [hunk 4]
```diff
@@ -51,8 +51,8 @@ public GenericSearchResult<Coupon> SearchCoupons(CouponSearchCriteria criteria)
                 {
                     sortInfos = new[] { new SortInfo { SortColumn = ReflectionUtility.GetPropertyName<Coupon>(x => x.Code), SortDirection = SortDirection.Descending } };
                 }
-                query = query.OrderBySortInfos(sortInfos);
-                
+                query = query.OrderBySortInfos(sortInfos).ThenBy(x => x.Id);
```
Comment (tatarincev) [inline VirtoCommerce.MarketingModule.Data/Services/CouponService.cs]: GetByIds(ids); below doesn't preserve sorting order.

File: VirtoCommerce.MarketingModule.Data/Services/MarketingSearchServiceImpl.cs [hunk 5]
```diff
@@ -98,7 +98,7 @@ public GenericSearchResult<coreModel.DynamicContentItem> SearchContentItems(Dyna
                 {
                     sortInfos = new[] { new SortInfo { SortColumn = ReflectionUtility.GetPropertyName<coreModel.DynamicContentItem>(x => x.Name), SortDirection = SortDirection.Ascending } };
                 }
-                query = query.OrderBySortInfos(sortInfos);
+                query = query.OrderBySortInfos(sortInfos).ThenBy(x => x.Id);
```
Comment (tatarincev) [inline VirtoCommerce.MarketingModule.Data/Services/MarketingSearchServiceImpl.cs]: GetByIds(ids); below doesn't preserve sorting order.

File: VirtoCommerce.MarketingModule.Data/Services/MarketingSearchServiceImpl.cs [hunk 6]
```diff
@@ -129,7 +129,7 @@ public GenericSearchResult<coreModel.DynamicContentPlace> SearchContentPlaces(Dy
                 {
                     sortInfos = new[] { new SortInfo { SortColumn = ReflectionUtility.GetPropertyName<coreModel.DynamicContentPlace>(x => x.Name), SortDirection = SortDirection.Ascending } };
                 }
-                query = query.OrderBySortInfos(sortInfos);
+                query = query.OrderBySortInfos(sortInfos).ThenBy(x => x.Id);
```
Comment (tatarincev) [inline VirtoCommerce.MarketingModule.Data/Services/MarketingSearchServiceImpl.cs]: GetByIds(ids); below doesn't preserve sorting order.

File: VirtoCommerce.MarketingModule.Data/Services/MarketingSearchServiceImpl.cs [hunk 7]
```diff
@@ -163,7 +163,7 @@ public GenericSearchResult<coreModel.DynamicContentPublication> SearchContentPub
                 {
                     sortInfos = new[] { new SortInfo { SortColumn = ReflectionUtility.GetPropertyName<coreModel.DynamicContentPublication>(x => x.Name), SortDirection = SortDirection.Ascending } };
                 }
-                query = query.OrderBySortInfos(sortInfos);
+                query = query.OrderBySortInfos(sortInfos).ThenBy(x => x.Id);
```
Comment (tatarincev) [inline VirtoCommerce.MarketingModule.Data/Services/MarketingSearchServiceImpl.cs]: GetByIds(ids); below doesn't preserve sorting order.

File: VirtoCommerce.MarketingModule.Data/Services/CouponService.cs (hunk 4 above)
Comment (AliveMen) [reply VirtoCommerce.MarketingModule.Data/Services/CouponService.cs]: fixed

File: VirtoCommerce.MarketingModule.Data/Services/MarketingSearchServiceImpl.cs (hunk 5 above)
Comment (AliveMen) [reply VirtoCommerce.MarketingModule.Data/Services/MarketingSearchServiceImpl.cs]: fixed

File: VirtoCommerce.MarketingModule.Data/Services/MarketingSearchServiceImpl.cs (hunk 6 above)
Comment (AliveMen) [reply VirtoCommerce.MarketingModule.Data/Services/MarketingSearchServiceImpl.cs]: fixed

File: VirtoCommerce.MarketingModule.Data/Services/MarketingSearchServiceImpl.cs (hunk 7 above)
Comment (AliveMen) [reply VirtoCommerce.MarketingModule.Data/Services/MarketingSearchServiceImpl.cs]: fixed

File: VirtoCommerce.MarketingModule.Data/Services/CouponService.cs [hunk 8]
```diff
@@ -51,15 +51,15 @@ public GenericSearchResult<Coupon> SearchCoupons(CouponSearchCriteria criteria)
...
+
                 var searchResult = new GenericSearchResult<Coupon> { TotalCount = query.Count() };
 
                 var ids = query.Select(x => x.Id)
                                .Skip(criteria.Skip)
-                               .Take(criteria.Take).ToArray();
+                               .Take(criteria.Take).ToList();
```
Comment (artem-dudarev) [inline VirtoCommerce.MarketingModule.Data/Services/CouponService.cs]: Move  `.ToLIst();` to next line

File: VirtoCommerce.MarketingModule.Data/Services/CouponService.cs [hunk 9]
```diff
@@ -51,15 +51,15 @@ public GenericSearchResult<Coupon> SearchCoupons(CouponSearchCriteria criteria)
...
 
-                searchResult.Results = GetByIds(ids);
+                searchResult.Results = GetByIds(ids.ToArray()).OrderBy(x => ids.IndexOf(x.Id)).ToList(); ;
```
Comment (artem-dudarev) [inline VirtoCommerce.MarketingModule.Data/Services/CouponService.cs]: Split into multiple lines.
The same for other files.

File: VirtoCommerce.MarketingModule.Data/Services/CouponService.cs (hunk 8 above)
Comment (AliveMen) [reply VirtoCommerce.MarketingModule.Data/Services/CouponService.cs]: fixed

File: VirtoCommerce.MarketingModule.Data/Services/CouponService.cs (hunk 9 above)
Comment (AliveMen) [reply VirtoCommerce.MarketingModule.Data/Services/CouponService.cs]: fixed

File: .gitignore [hunk 10]
```diff
@@ -156,3 +155,13 @@ $RECYCLE.BIN/
 *.zip
 
 .vs/
+/.idea/.idea.VirtoCommerce.MarketingModule/.idea/.name
```
Comment (yecli) [inline .gitignore]: could one record like /.idea/ be used?

Comment (AliveMen) [reply .gitignore]: fixed

File: .gitignore [hunk 11]
```diff
@@ -156,3 +155,4 @@ $RECYCLE.BIN/
 *.zip
 
 .vs/
+/.idea/*
```
[the remark this answers is not in the document]
Comment (AliveMen) [reply .gitignore]: fixed

File: VirtoCommerce.MarketingModule.Data/Services/CouponService.cs [hunk 12]
```diff
@@ -51,15 +51,18 @@ public GenericSearchResult<Coupon> SearchCoupons(CouponSearchCriteria criteria)
...
-                
+                query = query.OrderBySortInfos(sortInfos).ThenBy(x => x.Id);
+
                 var searchResult = new GenericSearchResult<Coupon> { TotalCount = query.Count() };
 
                 var ids = query.Select(x => x.Id)
                                .Skip(criteria.Skip)
-                               .Take(criteria.Take).ToArray();
+                               .Take(criteria.Take)
+                               .ToList();
```
[the remark this answers is not in the document]
Comment (AliveMen) [reply VirtoCommerce.MarketingModule.Data/Services/CouponService.cs]: fixed

File: VirtoCommerce.MarketingModule.Data/Services/MarketingSearchServiceImpl.cs [hunk 13]
```diff
@@ -36,14 +36,18 @@ public virtual GenericSearchResult<coreModel.Promotion> SearchPromotions(Promoti
...
 
                 var ids = query.Select(x => x.Id)
                                .Skip(criteria.Skip)
-                               .Take(criteria.Take).ToArray();
-                retVal.Results = _promotionSrevice.GetPromotionsByIds(ids).OrderBy(p => ids.ToList().IndexOf(p.Id)).ToList();
+                               .Take(criteria.Take)
+                               .ToArray();
+
+                retVal.Results = _promotionSrevice.GetPromotionsByIds(ids)
+                    .OrderBy(p => ids.ToList().IndexOf(p.Id))
```
[the remark this answers is not in the document]
Comment (AliveMen) [reply VirtoCommerce.MarketingModule.Data/Services/MarketingSearchServiceImpl.cs]: fixed

File: VirtoCommerce.MarketingModule.Data/Services/MarketingSearchServiceImpl.cs [hunk 14]
```diff
@@ -98,14 +102,18 @@ public GenericSearchResult<coreModel.DynamicContentItem> SearchContentItems(Dyna
...
+                query = query.OrderBySortInfos(sortInfos).ThenBy(x => x.Id);
 
                 retVal.TotalCount = query.Count();
 
                 var ids = query.Select(x => x.Id)
                                .Skip(criteria.Skip)
-                               .Take(criteria.Take).ToArray();
-                retVal.Results = _dynamicContentService.GetContentItemsByIds(ids);
+                               .Take(criteria.Take)
+                               .ToList();
```
[the remark this answers is not in the document]
Comment (AliveMen) [reply VirtoCommerce.MarketingModule.Data/Services/MarketingSearchServiceImpl.cs]: fixed

File: VirtoCommerce.MarketingModule.Data/Services/MarketingSearchServiceImpl.cs [hunk 15]
```diff
@@ -129,13 +137,17 @@ public GenericSearchResult<coreModel.DynamicContentPlace> SearchContentPlaces(Dy
...
+                query = query.OrderBySortInfos(sortInfos).ThenBy(x => x.Id);
 
                 retVal.TotalCount = query.Count();
                 var ids = query.Select(x => x.Id)
-                               .Skip(criteria.Skip)
-                               .Take(criteria.Take).ToArray();
-                retVal.Results = _dynamicContentService.GetPlacesByIds(ids);
+                    .Skip(criteria.Skip)
+                    .Take(criteria.Take)
+                    .ToList();
```
[the remark this answers is not in the document]
Comment (AliveMen) [reply VirtoCommerce.MarketingModule.Data/Services/MarketingSearchServiceImpl.cs]: fixed

File: VirtoCommerce.MarketingModule.Data/Services/MarketingSearchServiceImpl.cs [hunk 16]
```diff
@@ -163,14 +175,18 @@ public GenericSearchResult<coreModel.DynamicContentPublication> SearchContentPub
...
 
                 retVal.TotalCount = query.Count();
 
                 var ids = query.Select(x => x.Id)
-                           .Skip(criteria.Skip)
-                           .Take(criteria.Take).ToArray();
-                retVal.Results = _dynamicContentService.GetPublicationsByIds(ids);
+                    .Skip(criteria.Skip)
+                    .Take(criteria.Take)
+                    .ToList();
```
[the remark this answers is not in the document]
Comment (AliveMen) [reply VirtoCommerce.MarketingModule.Data/Services/MarketingSearchServiceImpl.cs]: fixed

File: VirtoCommerce.MarketingModule.Data/Services/MarketingSearchServiceImpl.cs [hunk 17]
```diff
@@ -98,14 +102,18 @@ public GenericSearchResult<coreModel.DynamicContentItem> SearchContentItems(Dyna
...
                 retVal.TotalCount = query.Count();
 
                 var ids = query.Select(x => x.Id)
                                .Skip(criteria.Skip)
-                               .Take(criteria.Take).ToArray();
-                retVal.Results = _dynamicContentService.GetContentItemsByIds(ids);
+                               .Take(criteria.Take)
+                               .ToArray();
+
+                retVal.Results = _dynamicContentService.GetContentItemsByIds(ids.ToArray())
```
Comment (yecli) [inline VirtoCommerce.MarketingModule.Data/Services/MarketingSearchServiceImpl.cs]: No need in additional .ToArray()

File: VirtoCommerce.MarketingModule.Data/Services/MarketingSearchServiceImpl.cs [hunk 18]
```diff
@@ -129,13 +137,17 @@ public GenericSearchResult<coreModel.DynamicContentPlace> SearchContentPlaces(Dy
...
                 retVal.TotalCount = query.Count();
                 var ids = query.Select(x => x.Id)
-                               .Skip(criteria.Skip)
-                               .Take(criteria.Take).ToArray();
-                retVal.Results = _dynamicContentService.GetPlacesByIds(ids);
+                    .Skip(criteria.Skip)
+                    .Take(criteria.Take)
+                    .ToArray();
+
+                retVal.Results = _dynamicContentService.GetPlacesByIds(ids.ToArray())
```
Comment (yecli) [inline VirtoCommerce.MarketingModule.Data/Services/MarketingSearchServiceImpl.cs]: No need in additional .ToArray()

File: VirtoCommerce.MarketingModule.Data/Services/MarketingSearchServiceImpl.cs [hunk 19]
```diff
@@ -163,14 +175,18 @@ public GenericSearchResult<coreModel.DynamicContentPublication> SearchContentPub
...
 
                 var ids = query.Select(x => x.Id)
-                           .Skip(criteria.Skip)
-                           .Take(criteria.Take).ToArray();
-                retVal.Results = _dynamicContentService.GetPublicationsByIds(ids);
+                    .Skip(criteria.Skip)
+                    .Take(criteria.Take)
+                    .ToArray();
+
+                retVal.Results = _dynamicContentService.GetPublicationsByIds(ids.ToArray())
```
Comment (yecli) [inline VirtoCommerce.MarketingModule.Data/Services/MarketingSearchServiceImpl.cs]: No need in additional .ToArray()

File: VirtoCommerce.MarketingModule.Data/Services/MarketingSearchServiceImpl.cs (hunk 18 above)
Comment (AliveMen) [reply VirtoCommerce.MarketingModule.Data/Services/MarketingSearchServiceImpl.cs]: fixed

File: VirtoCommerce.MarketingModule.Data/Services/MarketingSearchServiceImpl.cs (hunk 17 above)
Comment (AliveMen) [reply VirtoCommerce.MarketingModule.Data/Services/MarketingSearchServiceImpl.cs]: fixed

File: VirtoCommerce.MarketingModule.Data/Services/MarketingSearchServiceImpl.cs (hunk 19 above)
Comment (AliveMen) [reply VirtoCommerce.MarketingModule.Data/Services/MarketingSearchServiceImpl.cs]: fixed

**Pull request merged**